### ISSUE TITLE
Kokkos::subview() not working with aligned Views on HIP platforms

### ISSUE THREAD

**Issue by djglaze** [kokkos/kokkos#5862]
We have discovered that `Kokkos::subview()` fails with a device-side assert on HIP platforms if the source `Kokkos::View` has the `Kokkos::Aligned` memory trait.  Here is a reproducer for the error:

```
using data_type = double*[5];
using layout = Kokkos::LayoutLeft;
using memory_space = Kokkos::DefaultExecutionSpace::memory_space;
using memory_traits = Kokkos::MemoryTraits<Kokkos::Restrict | Kokkos::Aligned>;

using CustomView = Kokkos::View<data_type, layout, memory_space, memory_traits>;

TEST(defaultdevicetype, development_test) {

  CustomView customView("customView", 10);

  Kokkos::parallel_for("reproducer", customView.extent(0),
    KOKKOS_LAMBDA(int index) {
      const auto customSubView = Kokkos::subview(customView, index, Kokkos::ALL());
    });
}
```
and this is the error when running on a HIP platform (Caraway and Crusher, specifically):
```
:0: : Device-side assertion `Assigning NonAligned View or Pointer to Kokkos::View with Aligned attribute' failed.
:0: : Device-side assertion `Assigning NonAligned View or Pointer to Kokkos::View with Aligned attribute' failed.
:0: : Device-side assertion `Assigning NonAligned View or Pointer to Kokkos::View with Aligned attribute' failed.
:0: : Device-side assertion `Assigning NonAligned View or Pointer to Kokkos::View with Aligned attribute' failed.
:0: : Device-side assertion `Assigning NonAligned View or Pointer to Kokkos::View with Aligned attribute' failed.
:0: : Device-side assertion `Assigning NonAligned View or Pointer to Kokkos::View with Aligned attribute' failed.
:0: : Device-side assertion `Assigning NonAligned View or Pointer to Kokkos::View with Aligned attribute' failed.
:0: : Device-side assertion `Assigning NonAligned View or Pointer to Kokkos::View with Aligned attribute' failed.
:0:rocdevice.cpp            :2652: 3973238735976 us: 502330: [tid:0x7fc0e05a7700] Device::callbackQueue aborting with error : HSA_STATUS_ERROR_EXCEPTION: An HSAIL operation resulted in a hardware exception. code: 0x1016
Aborted (core dumped)
```
This is where the problem occurs:
https://github.com/kokkos/kokkos/blob/a45165d0954ba03fd7ac59e15bfb39a17d0e5dea/core/src/impl/Kokkos_ViewMapping.hpp#L2710-L2721
and it appears that we simply need to add a HIP guard next to the CUDA guard.
```
#ifdef KOKKOS_ENABLE_HIP
                      && (!(std::is_same<typename Traits::memory_space,
                                         Kokkos::Experimental::HIPSpace>::value ||
                            std::is_same<typename Traits::memory_space,
                                         Kokkos::Experimental::HIPManagedSpace>::value))
#endif
```
There appears to be three different versions of the `ViewDataHandle` class in this file that need the same treatment.

**Comment by masterleinad** [comment]
This also fails for `HostSpace`. It's not clear to me if taking a subview from a `Kokkos::Aligned` view should be `Kokkos::Aligned` as well but if so we should rather remove the check altogether.

**Comment by ajpowelsnl** [comment]
@djglaze -- have we resolved your subviews issues?

**Comment by masterleinad** [comment]
Fixed in #5878.
